### PR TITLE
Saas 6976 - Only allow deployment errors of type SaltoError or SaltoElementError

### PR DESCRIPTION
**Pull request by benisraeladi**
Alter the way we return errors from deployments, so that instead of returning one bulk error, we return multiple errors with their errors and ElemIDs (if relevant)

---

This is the beginnings of a PR to separate errors out in deployment. The way to look at this is as so:
commit 1: changes to core, adapter-api, and adapter-utils to allow this change
commit 2: Salesforce specific changes to make the adapter work

---

_Release Notes_: 
Core:
- Receive ElemIDs in deploy errors related to specific elements
Salesforce Adapter:
- Change deploy errors from `Error` type to `SaltoError` or `SaltoElementError`
---

_User Notifications_: _None_

### CODE REVIEW

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 1]
```diff
@@ -121,11 +123,15 @@ export const deployActions = async (
           log.warn(
             'Failed to deploy %s, errors: %s',
             item.groupKey,
-            result.errors.map(err => err.stack ?? err.message).join('\n\n'),
-          )
-          throw new Error(
-            `Failed to ${checkOnly ? 'validate' : 'deploy'} ${item.groupKey} with errors:\n${result.errors.join('\n')}`
+            // result.errors.map(err => err.stack ?? err.message).join('\n\n'),
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:126]: ?

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 2]
```diff
@@ -60,8 +60,10 @@ const deployAction = (
 }
 
 export class DeployError extends Error {
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:62]: Is there a reason not to have Severity here and make this extend SaltoError?

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 3]
```diff
@@ -121,11 +123,15 @@ export const deployActions = async (
...
-            result.errors.map(err => err.stack ?? err.message).join('\n\n'),
-          )
-          throw new Error(
-            `Failed to ${checkOnly ? 'validate' : 'deploy'} ${item.groupKey} with errors:\n${result.errors.join('\n')}`
+            // result.errors.map(err => err.stack ?? err.message).join('\n\n'),
+            result.errors.map(err => err.message).join('\n\n'),
           )
+          const modifiedErrors: DeployError[] = []
+          result.errors.forEach(error => {
+            const deployError = isSaltoElementError(error) ? new DeployError(item.groupKey, error.message, error.elemID) : new DeployError(item.groupKey, error.message)
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:131]: Why is this needed? If you made elementId optional in `DeployError` it would just work to do `new DeployError(item.groupKey, error.message, error.elemID)` in both cases, wouldn't it?

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 4]
```diff
@@ -121,11 +123,15 @@ export const deployActions = async (
...
             item.groupKey,
-            result.errors.map(err => err.stack ?? err.message).join('\n\n'),
-          )
-          throw new Error(
-            `Failed to ${checkOnly ? 'validate' : 'deploy'} ${item.groupKey} with errors:\n${result.errors.join('\n')}`
+            // result.errors.map(err => err.stack ?? err.message).join('\n\n'),
+            result.errors.map(err => err.message).join('\n\n'),
           )
+          const modifiedErrors: DeployError[] = []
+          result.errors.forEach(error => {
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:130]: Why is this a forEach and not a map?

**Review by adamro — COMMENTED** [review]
still early in the review but leaving the comments for now

File: packages/dag/src/nodemap.ts [hunk 5]
```diff
@@ -71,21 +67,21 @@ export type NodeHandler = (id: NodeId) => void
...
 
-  set(nodeId: NodeId, value: Error, visited: Set<string> = new Set()): this {
+  set(nodeId: NodeId, errors: Error[], visited: Set<string> = new Set()): this {
     const idAsString = nodeId.toString()
     if (visited.has(idAsString)) {
       return this
     }
     visited.add(idAsString)
-    super.set(nodeId, value)
+    super.set(nodeId, errors)
```
Comment (adamro) [inline packages/dag/src/nodemap.ts:82]: Won't this override the errors even if this NodeId already had errors before? Is this really legit?

File: packages/adapter-api/src/error.ts [hunk 6]
```diff
@@ -29,7 +29,7 @@ export type SaltoElementError = SaltoError & {
     elemID: ElemID
 }
 
-export const isSaltoElementError = (error: SaltoError | SaltoElementError):
+export const isSaltoElementError = (error: SaltoError | SaltoElementError | Error):
```
[the remark this answers is not in the document]
Comment (benisraeladi) [reply packages/adapter-api/src/error.ts:32]: This change is only temporary, the only reason `Error` is here is so that this will still compile for adapters that have not been touched yet

File: packages/core/src/core/deploy/deploy_actions.ts (hunk 2 above)
Comment (benisraeladi) [reply packages/core/src/core/deploy/deploy_actions.ts:62]: The problem with `SaltoError` is that it itself does not extend `Error`, meaning that it would be problematic throwing these errors (I tried and got an eslint error, and I tend to trust those unless proven otherwise)

File: packages/dag/src/nodemap.ts [hunk 7]
```diff
@@ -71,21 +67,21 @@ export type NodeHandler = (id: NodeId) => void
...
 
-class WalkErrors<T> extends Map<NodeId, Error> {
+class WalkErrors<T> extends Map<NodeId, Error[]> {
   // eslint-disable-next-line no-use-before-define
   constructor(readonly nodeMap: DAG<T>) {
     super()
   }
 
-  set(nodeId: NodeId, value: Error, visited: Set<string> = new Set()): this {
+  set(nodeId: NodeId, errors: Error[], visited: Set<string> = new Set()): this {
```
Comment (adamro) [inline packages/dag/src/nodemap.ts:76]: All the calls to this function in the file still pass `Error` and not `Error[]` (cause it's in then and it's an any) :\

File: packages/core/src/core/deploy/deploy_actions.ts (hunk 2 above)
Comment (adamro) [reply packages/core/src/core/deploy/deploy_actions.ts:62]: If we always return them and not throw them, do they really need to extend Error tho?

File: packages/core/src/core/deploy/deploy_actions.ts (hunk 3 above)
Comment (benisraeladi) [reply packages/core/src/core/deploy/deploy_actions.ts:131]: We currently need to throw them because of `WalkErrors` and `WalkError` logic. Doing a deep-dive to decide on new format for error collection will also work

File: packages/adapter-utils/src/deploy.ts [hunk 8]
```diff
@@ -28,7 +28,7 @@ export const deployInstance = async (
   if (!isInstanceChange(change)) {
     return {
       appliedChanges: [],
-      errors: [new Error('Only instance changes supported')],
+      errors: [{ elemID: getChangeData(change).elemID, message: 'Only instance changes supported', severity: 'Error' }],
```
[the remark this answers is not in the document]
Comment (benisraeladi) [reply packages/adapter-utils/src/deploy.ts:31]: It loos like this file isn't used anywhere, I think I'll just delete it in a different PR and revert this change

File: .vscode/settings.json [hunk 9]
```diff
@@ -0,0 +1,6 @@
+{
```
[the remark this answers is not in the document]
Comment (benisraeladi) [reply .vscode/settings.json:1]: No, should've been in `.ignore`. Will remove

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 10]
```diff
@@ -32,36 +34,56 @@ type DeployOrValidateParams = {
   checkOnly: boolean
 }
 
-const deployOrValidate = (
+const deployOrValidate = async (
   { adapter, adapterName, opts, checkOnly }: DeployOrValidateParams
-): Promise<DeployResult> => {
+): Promise<AdapterDeployResult> => {
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:39]: I like this but I am not 100% this is legit :)
Let's discuss this f2f

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 11]
```diff
@@ -32,36 +34,56 @@ type DeployOrValidateParams = {
...
   if (!checkOnly) {
-    return adapter.deploy(opts)
+    const deployRes = await adapter.deploy(opts)
+    return {
+      appliedChanges: deployRes.appliedChanges,
+      extraProperties: deployRes.extraProperties,
+      errors: deployRes.errors.map(error => (
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:45]: Probably best to DRY this part and write a function

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 12]
```diff
@@ -32,36 +34,56 @@ type DeployOrValidateParams = {
...
-): Promise<DeployResult> => {
+): Promise<AdapterDeployResult> => {
   const changes = [...planItem.changes()]
   const adapterName = getChangeData(changes[0]).elemID.adapter
   const adapter = adapters[adapterName]
   if (!adapter) {
     throw new Error(`Missing adapter for ${adapterName}`)
   }
   const opts = { changeGroup: { groupID: planItem.groupKey, changes } }
+  // TODOADI wrap deployOrValidate to make sure that the correct errors are returned
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:77]: ?

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 13]
```diff
@@ -32,36 +34,56 @@ type DeployOrValidateParams = {
...
   return deployOrValidate({ adapter, adapterName, opts, checkOnly })
 }
 
 export class DeployError extends Error {
-  constructor(readonly elementId: string | string[], message: string) {
+  public readonly elemID?: ElemID
+  constructor(readonly groupId: string | string[], message: string, elementId?: ElemID) {
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:83]: I think most of our code base has `elemID` and not `elementId`. You kept this for a reason?

Also, is there a reason for `groupId` to be typed like this and not be specific array or not?

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 14]
```diff
@@ -94,6 +116,14 @@ const updatePlanElement = (item: PlanItem, appliedChanges: ReadonlyArray<Change>
     })
 }
 
+class WalkDeployError extends Error {
+  public errors: ReadonlyArray<SaltoElementError | SaltoError>
+  constructor(deployErrors: ReadonlyArray<SaltoElementError | SaltoError>) {
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:121]: I know some really hate that syntax but you can use the constructor `(public errors: ReadonlyArray<SaltoElementError | SaltoError>)` and avoid 2 lines :) (120 and 123)

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 15]
```diff
@@ -32,36 +34,56 @@ type DeployOrValidateParams = {
...
   return deployOrValidate({ adapter, adapterName, opts, checkOnly })
 }
 
 export class DeployError extends Error {
-  constructor(readonly elementId: string | string[], message: string) {
+  public readonly elemID?: ElemID
+  constructor(readonly groupId: string | string[], message: string, elementId?: ElemID) {
     super(message)
+    this.elemID = elementId
+    this.groupId = groupId
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:86]: this is redundant isn't it? (can also be done for `elementId`)

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 16]
```diff
@@ -94,6 +116,14 @@ const updatePlanElement = (item: PlanItem, appliedChanges: ReadonlyArray<Change>
...
 
+class WalkDeployError extends Error {
+  public errors: ReadonlyArray<SaltoElementError | SaltoError>
+  constructor(deployErrors: ReadonlyArray<SaltoElementError | SaltoError>) {
+    super()
+    this.errors = deployErrors
+  }
+}
+
 export const deployActions = async (
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:127]: I am wondering wether we want to keep `deployActions`'s return type (`DeployActionResult`) to include errors that actually extends `Error` or want a different, Salto defined type here as well
We anyway recreate everything here
This can be a `SaltoError | SaltoElementError` or we can define a `SaltoGroupError` that is the basic thing here and have ones with both Group and Element. Instead of this. not sure it adds anything or is very different, just a choice we can make cause we are changing this now and most of the Errors we return typed are `SaltoError` and not extension of `Error`

**Review by adamro — COMMENTED** [review]
Not finished but went through most of the parts with real logic. Left comments but I think this direction is generally great!

File: packages/core/src/core/deploy/deploy_actions.ts (hunk 15 above)
Comment (benisraeladi) [reply packages/core/src/core/deploy/deploy_actions.ts:86]: That's only if we always expect `elemID` to exist, which can't happen because there are errors that don't relate to a specific elemID

File: packages/core/src/core/deploy/deploy_actions.ts (hunk 13 above)
Comment (benisraeladi) [reply packages/core/src/core/deploy/deploy_actions.ts:83]: In this part of the code originally `elementId` was actually the group ID. I changed the name for consistency, but kept the signature because I don't know how it'll impact the unaltered adapters/errors. I'll change it to `elemID` now that it's the correct type

File: packages/core/src/core/deploy/deploy_actions.ts (hunk 12 above)
Comment (benisraeladi) [reply packages/core/src/core/deploy/deploy_actions.ts:77]: oops

File: packages/core/src/core/deploy/deploy_actions.ts (hunk 15 above)
Comment (benisraeladi) [reply packages/core/src/core/deploy/deploy_actions.ts:86]: Oh, I think I misunderstood before. It's redundant and I'll remove it

File: packages/core/src/core/deploy/deploy_actions.ts (hunk 16 above)
Comment (adamro) [reply packages/core/src/core/deploy/deploy_actions.ts:125]: FYI @ori-moisis let's discuss this f2f the 3 of us

File: packages/core/src/core/deploy/deploy_actions.ts (hunk 2 above)
Comment (benisraeladi) [reply packages/core/src/core/deploy/deploy_actions.ts:83]: We decided to change the API and make it so that `DeployError` doesn't extend `Error` and instead is `SaltoError | SaltoElementError` with added `groupId`

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 17]
```diff
@@ -32,23 +34,42 @@ type DeployOrValidateParams = {
   checkOnly: boolean
 }
 
-const deployOrValidate = (
+const extractErrors = (deployRes: DeployResult, opts: DeployOptions): ReadonlyArray<SaltoElementError | SaltoError> =>
+  (deployRes.errors.map(error => (
+    error instanceof Error
+      ? { message: error.message,
+        severity: 'Error' as SeverityLevel,
+        group: opts.changeGroup.groupID }
```
[the remark this answers is not in the document]
Comment (adamro) [reply packages/core/src/core/deploy/deploy_actions.ts:42]: We later said that the behaviour we do want is when the adapter throws an error we want to create an Error per element in the group, tied to the Element's elemID. And not allow the adapters to return Errors.

File: packages/core/src/core/deploy/deploy_actions.ts (hunk 16 above)
Comment (adamro) [reply packages/core/src/core/deploy/deploy_actions.ts:125]: Here we said we should go through with this change and return SaltoError or SaltoElementError but with groupId and not something that extends Error

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 18]
```diff
@@ -32,23 +34,65 @@ type DeployOrValidateParams = {
...
+  ))
+  )
+
+const deployOrValidate = async (
   { adapter, adapterName, opts, checkOnly }: DeployOrValidateParams
-): Promise<DeployResult> => {
+): Promise<AdapterDeployResult> => {
   if (!checkOnly) {
-    return adapter.deploy(opts)
+    try {
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:59]: nit: Not completely your fault cause it was this way but the lack of DRY became a bit worse here.
Maybe have a 
`const deployOrValidateFn = checkOnly ? adapter.validate : adapter.deploy`
and then have the same code cause it's practically identical?

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 19]
```diff
@@ -142,16 +188,35 @@ export const deployActions = async (
         const item = deployPlan.getItem(key) as PlanItem
         if (nodeError instanceof NodeSkippedError) {
           reportProgress(item, 'cancelled', deployPlan.getItem(nodeError.causingNode).groupKey)
-          deployErrors.push(new DeployError(item.groupKey, `Element ${key} was not deployed, as it depends on element ${nodeError.causingNode} which failed to deploy`))
+          deployErrors.push({
+            groupId: item.groupKey,
+            message: `Element ${key} was not deployed, as it depends on element ${nodeError.causingNode} which failed to deploy`,
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:193]: Shouldn't we also attach this to the Element's elemID?

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 20]
```diff
@@ -142,16 +188,35 @@ export const deployActions = async (
...
+            severity: 'Error' as SeverityLevel,
+          })
+        } else if (nodeError instanceof WalkDeployError) {
+          deployErrors.push(...nodeError.errors.map(
+            deployError => (isSaltoElementError(deployError)
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:198]: Doesn't just adding the `groupId` to either work?

File: packages/cli/src/formatter.ts [hunk 21]
```diff
@@ -348,7 +349,21 @@ export const cancelDeployOutput = (checkOnly: boolean): string => [
   emptyLine(),
 ].join('\n')
 
-export const deployPhaseEpilogue = (numChanges: number, numErrors: number, checkOnly: boolean): string => {
+export const deployErrorsOutput = (deployErrors: DeployError[]): string => {
+  const errorOutputLines = [
+    emptyLine(),
+  ]
+  if (deployErrors.length > 0) {
+    errorOutputLines.push(error('Errors:'))
```
Comment (adamro) [inline packages/cli/src/formatter.ts:357]: We are ignoring the concept of Severity here?

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts [hunk 22]
```diff
@@ -193,7 +202,7 @@ export const retryFlow = async (
   const { retryDelay, retryableFailures } = client.dataRetry
 
   let successes: InstanceElement[] = []
-  let errMsgs: string[] = []
+  let errMsgs: (SaltoElementError | SaltoError)[] = []
```
Comment (adamro) [inline packages/salesforce-adapter/src/custom_object_instances_deploy.ts:205]: name change to errors?

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts [hunk 23]
```diff
@@ -405,10 +414,12 @@ const deployModifyChanges = async (
...
-  const diffApiNameErrors = await awu(diffApiNameData).map(async data => new Error(`Failed to update as api name prev=${await apiName(
-    data.before
-  )} and new=${await apiName(data.after)} are different`)).toArray()
-  const errors = errorMessages.map(error => new Error(error)).concat(diffApiNameErrors)
+  const diffApiNameErrors: SaltoElementError[] = await awu(diffApiNameData).map(async data => ({
+    elemID: data.before.elemID,
+    message: `Failed to update as api name prev=${await apiName(data.before)} and new=${await apiName(data.after)} are different`,
+    severity: 'Error' as SeverityLevel,
+  })).toArray()
+  const errors: (SaltoElementError | SaltoError)[] = [...errorMessages, ...diffApiNameErrors]
```
Comment (adamro) [inline packages/salesforce-adapter/src/custom_object_instances_deploy.ts:422]: `errorMessages` sounds like a wrong name now as well

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts [hunk 24]
```diff
@@ -66,12 +69,18 @@ ${safeJsonStringify(instance.value, undefined, 2,)}
   })
 )
 
-const getErrorMessagesFromInstAndResults = (instancesAndResults: InstanceAndResult[]): string[] =>
+const getErrorMessagesFromInstAndResults = (instancesAndResults: InstanceAndResult[]): SaltoElementError[] =>
   instancesAndResults
-    .map(({ instance, result }) => `${instance.elemID.name}:
-    \t${result.errors?.join('\n\t')}`)
+    .flatMap(({ instance, result }) => {
```
Comment (adamro) [inline packages/salesforce-adapter/src/custom_object_instances_deploy.ts:74]: Our code base usually uses `filter` and then `map` instead of the if inside and the let

File: packages/salesforce-adapter/src/metadata_deploy.ts [hunk 25]
```diff
@@ -195,37 +198,39 @@ const isUnFoundDelete = (message: DeployMessage, deletionsPackageName: string):
...
+      ),
+      severity: 'Error' as SeverityLevel,
+    }))
   const componentErrors = allFailureMessages
     .filter(failure => !isUnFoundDelete(failure, deletionsPackageName))
     .map(getUserFriendlyDeployMessage)
-    .map(failure => new Error(
-      `Failed to ${checkOnly ? 'validate' : 'deploy'} ${failure.fullName} with error: ${failure.problem} (${failure.problemType})`
+    .map(failure => (
+      { elemID: ElemID.fromFullName(failure.fullName), message: `${failure.problem} (${failure.problemType})`, severity: 'Error' as SeverityLevel }
```
Comment (adamro) [inline packages/salesforce-adapter/src/metadata_deploy.ts:222]: How come we are removing the failed to XXX here?

**Review by adamro — COMMENTED** [review]
We are very close :D

**Review by adamro — COMMENTED** [review]
We are very close :D

File: packages/cli/src/formatter.ts (hunk 21 above)
Comment (benisraeladi) [reply packages/cli/src/formatter.ts:357]: Yes

Comment (adamro) [reply packages/cli/src/formatter.ts:357]: Why? :)
I thought there's already a use-case of Warning/Info we thought of and we know we will use it. Why not differentiate between the cases in the CLI from the go?

File: packages/adapter-api/src/adapter.ts [hunk 26]
```diff
@@ -41,12 +41,23 @@ export type DeployExtraProperties = {
   groups?: Group[]
 }
 
-export type DeployResult = {
+type SaltoDeployErrors = {
+  errors: ReadonlyArray<SaltoError | SaltoElementError>
+}
+
+type AdapterDeployErrors = {
```
[the remark this answers is not in the document]
Comment (benisraeladi) [reply packages/adapter-api/src/adapter.ts:48]: In the old scheme, adapters *do* return `Error` typed errors. The reason for `AdapterDeployErrors` is a gateway so that `deployOrValidate` can change all `Error` type errors to some sort of `SaltoError`. In the future, only `SaltoDeployErrors` will be used, and hopefully we will be able to delete `AdapterDeployErrors`

File: packages/salesforce-adapter/src/metadata_deploy.ts (hunk 25 above)
Comment (benisraeladi) [reply packages/salesforce-adapter/src/metadata_deploy.ts:222]: It felt redundant to me, but if you think it should stay I'll bring it back

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 27]
```diff
@@ -33,23 +35,53 @@ type DeployOrValidateParams = {
...
+      severity: 'Error' as SeverityLevel,
+      elemID: getChangeData(change).elemID }
+  ))
+  )
+
+const extractErrors = (
+  result: DeployResult,
+  changes: readonly Change<ChangeDataType>[]
+): ReadonlyArray<SaltoElementError | SaltoError> =>
+  result.errors.map(error =>
```
Comment (adamro) [inline packages/core/src/core/deploy/deploy_actions.ts:52]: nit: result.errors.flatMap(...

**Review by adamro — COMMENTED** [review]
I think we just need to decide if to keep the adapter API as is with the ability to return Error or not (not ---> Adding the code that transforms it to a SaltoElementError per Element to adapter-component/utils and have all the adapters use it instead of returning Error for now)

File: packages/core/src/core/deploy/deploy_actions.ts [hunk 28]
```diff
@@ -33,23 +35,53 @@ type DeployOrValidateParams = {
...
-): Promise<DeployResult> => {
-  if (!checkOnly) {
-    return adapter.deploy(opts)
-  }
-  if (_.isUndefined(adapter.validate)) {
-    throw new Error(`Check-Only deployment is not supported in adapter ${adapterName}`)
+): Promise<AdapterDeployResult> => {
+  try {
+    const deployOrValidateFn = checkOnly ? adapter.validate : adapter.deploy
+    if (_.isUndefined(deployOrValidateFn)) {
```
[the remark this answers is not in the document]
Comment (benisraeladi) [reply packages/core/src/core/deploy/deploy_actions.ts:63]: That's what was used in the function beforehand. Would the right alternative be deployOrValidateFn === undefined?

[the remark this answers is not in the document]
Comment (adamro) [reply packages/core/src/core/deploy/deploy_actions.ts:63]: yep

**Review by adamro — APPROVED** [review]
🔥 
Great work on this! Surely not the most trivial area of our code and I think the end result is a huge jump in the right directly. Well executed!

**Review by adamro — APPROVED** [review]
🔥 
Great work on this! Surely not the most trivial area of our code and I think the end result is a huge jump in the right directly. Well executed!
